screen orientation should match other browsers

marcoscaceres opened the pull request:
#### 8df6fffd2980f76772db9727294efa1e91335a32
<pre>
screen orientation should match other browsers
<a href="https://bugs.webkit.org/show_bug.cgi?id=259762">https://bugs.webkit.org/show_bug.cgi?id=259762</a>
rdar://113312337

Reviewed by NOBODY (OOPS!).

Flipped the order so it matches what browsers on Android report, that is:

  - Device turned clockwise is &quot;landscape-secondary&quot;, angle 270.
  - Device turned counter-clockwise, &quot;landscape-primary&quot;, angle 90.
  - iOS actual legacy angles are the opposite of Android.

* LayoutTests/fast/screen-orientation/orientation-in-resize-event.html:
* LayoutTests/fast/screen-orientation/orientation-types-and-angles-expected.txt: Added.
* LayoutTests/fast/screen-orientation/orientation-types-and-angles.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::orientation const):
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::toUIInterfaceOrientationMask):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::toScreenOrientationType):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::lockScreenOrientation):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8df6fffd2980f76772db9727294efa1e91335a32

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15236 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16328 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14717 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15116 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16426 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17059 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12559 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13150 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13638 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13316 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16553 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13867 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11830 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13161 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13007 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17496 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/ScreenOrientation.cpp:
```diff
@@ -184,7 +184,7 @@ uint16_t ScreenOrientation::angle() const
     auto orientation = manager ? manager->currentOrientation() : naturalScreenOrientationType();
 
-    // https://w3c.github.io/screen-orientation/#dfn-screen-orientation-values-table
+    // https://w3c.github.io/screen-orientation/#the-current-screen-orientation-type-and-angle
```
Doesn't this function in WebPageProxyIOS.mm also need to be fixed?
```
WebCore::ScreenOrientationType WebPageProxy::toScreenOrientationType(IntDegrees angle)
{
    if (angle == -90)
        return WebCore::ScreenOrientationType::LandscapeSecondary;
    if (angle == 180)
        return WebCore::ScreenOrientationType::PortraitSecondary;
    if (angle == 90)
        return WebCore::ScreenOrientationType::LandscapePrimary;
    return WebCore::ScreenOrientationType::PortraitPrimary;
}
```

marcoscaceres replied:
The above looks right to me... The above matches what Firefox and Chrome report.

marcoscaceres replied:
Ok, yeah... I thought the angles were the same on Android and iOS... but they are flipped 🫠

cdumez reviewed (changes requested):
The changes to TestController are only used for testing and wouldn't impact in-browser behavior.

This only other behavior change you made is in PageClientImplIOS.mm's toUIInterfaceOrientationMask() and is only used for telling the OS which orientation we want to LOCK.

Unless I am missing anything, this patch does NOT actually impact the orientation / angle being reported to the page when you rotate your phone. For this, you would have needed to modify the toScreenOrientationType() function in WebPageProxyIOS.mm.

Since you say that we match other browsers with this patch, I assume we match them without this patch as well?

You might be fixing a bug with regards to *locking* a particular orientation but there is no testing to cover that so it is hard to tell.

marcoscaceres committed:
screen orientation should match other browsers
https://bugs.webkit.org/show_bug.cgi?id=259762
rdar://113312337

Reviewed by NOBODY (OOPS!).

Flipped the order so it matches what browsers on Android report, that is:

  - Device turned clockwise is "landscape-secondary", angle 270.
  - Device turned counter-clockwise, "landscape-primary", angle 90.
  - iOS actual legacy angles are the opposite of Android.

* LayoutTests/fast/screen-orientation/orientation-in-resize-event.html:
* LayoutTests/fast/screen-orientation/orientation-types-and-angles-expected.txt: Added.
* LayoutTests/fast/screen-orientation/orientation-types-and-angles.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::orientation const):
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::toUIInterfaceOrientationMask):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::toScreenOrientationType):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::lockScreenOrientation):

cdumez commented on LayoutTests/fast/screen-orientation/orientation-types-and-angles.html:
```diff
@@ -0,0 +1,111 @@
+                    await runRotationUIScript(rotation);
+
+                    const { type, angle, legacy } = rotations.get(rotation);
+                    debug(
```
We don't need this.

cdumez commented on LayoutTests/fast/screen-orientation/orientation-types-and-angles.html:
```diff
@@ -0,0 +1,111 @@
+                    debug(
+                        `Expected angle: ${angle} | actual: ${screen.orientation.angle}`
+                    );
+                    angleIsCorrect = angle === screen.orientation.angle;
```
seems we should be using:
```
shouldBe("screen.orientation.angle", "angle");
```
or
```
shouldBe("screen.orientation.angle", "" + angle);
```
if the first former one doesn't print the angle value in the output.

cdumez commented on LayoutTests/fast/screen-orientation/orientation-types-and-angles.html:
```diff
@@ -0,0 +1,111 @@
+                    shouldBeTrue("angleIsCorrect");
+
+                    debug(
+                        `Expected type: ${type} | actual: ${screen.orientation.type}`
```
ditto.

cdumez commented on LayoutTests/fast/screen-orientation/orientation-types-and-angles.html:
```diff
@@ -0,0 +1,111 @@
+                    shouldBeTrue("typeIsCorrect");
+
+                    debug(
+                        `Expected window.orientation: ${legacy} | actual: ${window.orientation}`
```
ditto

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -376,7 +376,7 @@ void LocalFrame::orientationChanged()
 {
     if (auto* page = this->page())
-        return page->chrome().client().deviceOrientation();
+        return page->chrome().client().deviceOrientation() * -1;
```
I doubt we should be changing this. This is code pre-dating the screen orientation API.

marcoscaceres replied:
Right, but we report the opposite angles :( Not sure what to do here.

marcoscaceres replied:
FWIW, it's reporting angles the same as before the code changes (from my testing)... so they would need to be flipped somewhere if not here?

cdumez replied:
You say "angle" and I am confused because there is:
(1) window.orientation, which is legacy and for which we shouldn't change behavior.
(2) window.screen.orientation.angle, which is the new API and which value is deduced from window.screen.orientation.type

I don't think we should change the values returned by (1), even if they are reversed compared to Android. Since this is long-standing behavior, I believe backward compatibility is more important. As far as I know, the code you are modifying in this function should only impact (1), which is why I am surprised.

cdumez commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -316,11 +316,11 @@ static inline float adjustedUnexposedMaxEdge(float documentEdge, float exposedRe
 {
     if (angle == -90)
-        return WebCore::ScreenOrientationType::LandscapeSecondary;
+        return WebCore::ScreenOrientationType::LandscapePrimary;
```
I don't understand, you had said earlier that our orientation.type matched other browsers on trunk already. However, you are now reversing it here?

marcoscaceres replied:
Right, yeah... let me see if undoing this and not changing the window.orientation yields the same results.
